src/m365/spfx/commands/project/project-azuredevops-pipeline-add.ts

```diff
@@ -0,0 +1,213 @@
+          return `${args.options.loginMethod} is not a valid login method. Allowed values are ${SpfxProjectAzureDevOpsPipelineAddCommand.loginMethod.join(', ')}`;
+        }
+
+        if (args.options.scope && SpfxProjectAzureDevOpsPipelineAddCommand.scope.indexOf(args.options.scope) < 0) {
+          return `${args.options.scope} is not a valid scope. Allowed values are ${SpfxProjectAzureDevOpsPipelineAddCommand.scope.join(', ')}`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    this.projectRootPath = this.getProjectRoot(process.cwd());
+    if (this.projectRootPath === null) {
+      throw new CommandError(`Couldn't find project root folder`, SpfxProjectAzureDevOpsPipelineAddCommand.ERROR_NO_PROJECT_ROOT_FOLDER);
+    }
+
+    const solutionPackageJsonFile: string = path.join(this.projectRootPath, 'package.json');
+    const packageJson: string = fs.readFileSync(solutionPackageJsonFile, 'utf-8');
+    const solutionName = JSON.parse(packageJson).name;
+
+    if (this.debug) {
+      logger.logToStderr(`Adding Azure DevOps pipeline in the current SPFx project`);
+    }
+
+    try {
+      this.updatePipeline(solutionName, pipeline, args.options);
+      this.savePipeline(pipeline);
+    }
+    catch (error: any) {
+      throw new CommandError(error);
+    }
+  }
+
+  private savePipeline(pipeline: azureDevOpsPipeline): void {
+    const azureDevOpsPath: string = path.join(this.projectRootPath as string, '.azuredevops');
+    fsUtil.ensureDirectory(azureDevOpsPath);
+
+    const pipelinesPath: string = path.join(azureDevOpsPath, 'pipelines');
+    fsUtil.ensureDirectory(pipelinesPath);
+
+    const pipelineFile: string = path.join(pipelinesPath, 'deploy-spfx-solution.yml');
+    fs.writeFileSync(path.resolve(pipelineFile), yaml.stringify(pipeline), 'utf-8');
+  }
+
+  private updatePipeline(solutionName: string, pipeline: azureDevOpsPipeline, options: GlobalOptions): void {
+    if (options.name) {
+      pipeline.name = options.name;
+    }
+    else {
+      delete pipeline.name;
+    }
+
+    if (options.branchName) {
+      pipeline.trigger.branches.include[0] = options.branchName;
+    }
+
+    const version = this.getProjectVersion();
+
+    if (!version) {
+      throw `Unable to determine the version of the current SharePoint Framework project`;
+    }
+
+    const match = version.match(/^(\d+)\.(\d+)\.(\d+)$/);
+    const minorVersionString = match ? match[2] : null;
+    const minorVersion = minorVersionString ? Number(minorVersionString) : null;
+
+    if (minorVersion === null || isNaN(minorVersion)) {
+      throw `Unable to determine the minor version of the current SharePoint Framework project`;
+    }
+
+    if (minorVersion < 18) {
+      this.getNodeAction(pipeline).inputs!.versionSpec = '16.x';
```

Comment:
If we suppress inputs as non-nullable, why do we define them as nullable in our type in the first place?

Reply:
good point. Each task in ADO has `inputs` so it does not need to be nullable in the interface 👍

Reply:
We may have a task without `inputs` option defined so it has to be nullable in the type.
Here we assume we take a specific action/task from the pipeline whic is the taks that specifies the node version and we 'know' that it has `inputs` as we control over the flow definition itself. Thats why I supress, we may do a `if` condition instead

Reply:
If the task can be null, then we should do an `if`. On the other hand, we don't accept external input for it right? So for us, the input is always filled in, correct?

Reply:
Exactly, this field is not a user input but it's based on our internal JSON template of the pipeline. The user may not pass but only modify it's values

Reply:
Got it, so if the data is only coming from us, let's modify the type to our needs to avoid unnecessary checks

Reply:
@waldekmastykarz sorry I am not sure if I follow 😟. How should I modify the type to avoid checks? IMO the current type is how it should be. 
Let's go over the following example. In the ADO pipeline we may have something like
```
steps:
          ......
          - task: Npm@1
            displayName: Install CLI for Microsoft 365
            inputs:
              command: custom
              verbose: false
              customCommand: install -g @pnp/cli-microsoft365
          - script: >
             someScriptHere
            displayName: CLI for Microsoft 365 Deploy App
```

as a result we have the following JSON object in our code which we maintain in our code
```
steps: [
            {
              task: "Npm@1",
              displayName: "Install CLI for Microsoft 365",
              inputs: {
                command: "custom",
                verbose: false,
                customCommand: "install -g @pnp/cli-microsoft365"
              }
            },
            {
              script: "someScriptHere",
              displayName: "CLI for Microsoft 365 Deploy App"
            }
          ]
```

So now if we want to create a type for that which will cover both cases we need to say that `steps` will be an array of some kind of object that will support either a task or script case. 
So something like:

```
...
 steps: AzureDevOpsPipelineStep[];
....

export interface AzureDevOpsPipelineStep {
  task?: string;
  script?: string;
  displayName?: string;
  inputs?: {
    command?: string;
    verbose?: boolean;
    customCommand?: string;
  }
}
```

in the above case the `inputs` property has to be nullable as we may have a `step` which may not have this property so in order to have a universal type that support all scenarios most properties have to be like that.
Am I missing something here? How should I correct this to avoid making the addtional checks?

Reply:
Thanks for clarifying. If it has to be null, then we should check for null value in our code to avoid making assumptions that might catch us off guard.

Reply:
ok done 👍. Thanks for your patience 🙂